### PR TITLE
Mark failed documents

### PR DESCRIPTION
**Pull request by elena-vi**
Screenshot 2020-01-15 at 09 03 50

Screenshot 2020-01-15 at 09 26 25

Screenshot 2020-01-15 at 09 26 44

### CODE REVIEW

File: app/controllers/documents_controller.rb [hunk 1]
```diff
@@ -26,6 +26,14 @@ def index
     @documents = use_cases.get_all_documents.execute(filters: filters_param)
   end
 
+  def review_failure
+    @document = use_cases.review_document_failure.execute(document_id: params.require(:id))
+
+    flash[:notice] = response.code.to_i == 200 ? 'Successfully marked as reviewed' : "Unable to mark as reviewed"
```
Comment (richrace) [inline app/controllers/documents_controller.rb]: I think the gateway will throw a `Exceptions::IncomeApiError::NotFoundError` error and not get the response.

File: lib/hackney/income/documents_gateway.rb [hunk 2]
```diff
@@ -34,6 +34,19 @@ def get_all(filters: {})
...
 
+      def review_failure(document_id:)
+        uri = URI("#{@api_host}#{DOCUMENTS_ENDPOINT}#{document_id}/review_failure")
+        req = Net::HTTP::Patch.new(uri)
+        req['X-Api-Key'] = @api_key
+
+        res = Net::HTTP.start(uri.hostname, uri.port, use_ssl: (uri.scheme == 'https')) { |http| http.request(req) }
+
+        unless res.is_a? Net::HTTPSuccess
+          raise Exceptions::IncomeApiError::NotFoundError.new(res), 'when trying to get all documents'
```
Comment (richrace) [inline lib/hackney/income/documents_gateway.rb]: `'when trying to get all documents'` Doesn't make sense here.

File: spec/controllers/documents_controller_spec.rb [hunk 3]
```diff
@@ -98,4 +98,31 @@
...
+          .to receive(:review_failure).with(document_id: id).and_return(document_response)
+
+        patch :index, params: {id: id}
+
+        expect(flash[:notice]).to eq('Unable to pause: Internal server error')
+      end
+    end
+
+    context 'when not successfully reviewed' do
+      let(:document_response) { Net::HTTPResponse.new(1.1, 400, 'NOT OK') }
```
Comment (richrace) [inline spec/controllers/documents_controller_spec.rb]: Curious to know why this isn't a triggering the `unless res.is_a? Net::HTTPSuccess` in the Gateway.

File: app/helpers/tenancy_helper.rb [hunk 4]
```diff
@@ -34,4 +34,13 @@ def select_all_other_outgoing_charges(group_summary_list)
   def calculate_sum_of_all_other_outgoing_charges(group_summary_list)
     group_summary_list.select { |i| i[:transaction] && i[:type] != BASIC_RENT_TRANSACTION_TYPE && i[:value].positive? }.sum { |i| i[:value] }
   end
+
+  def insert_document_links_to_action_diary(entry)
+    document_url = 'documents?payment_ref='
+    if entry.include?(document_url)
+      entry.gsub!(/#{Regexp.quote(document_url)}/, request.base_url+'/\0') # prepend environment url to document url
+      entry.gsub!(URI.regexp, '<a href="\0">\0</a>') # insert anchor tag
```
Comment (richrace) [inline app/helpers/tenancy_helper.rb]: Neat. But ugly. 🤷‍♂

File: app/controllers/documents_controller.rb [hunk 5]
```diff
@@ -26,6 +26,15 @@ def index
     @documents = use_cases.get_all_documents.execute(filters: filters_param)
   end
 
+  def review_failure
+    @document = use_cases.review_document_failure.execute(document_id: params.require(:id))
+    flash[:notice] = 'Successfully marked as reviewed'
+    redirect_back fallback_location: documents_path
+  rescue Exceptions::IncomeApiError::NotFoundError
+    flash[:notice] = "An error occurred while marking document #{params.require(:id)} as reviewed"
+    redirect_back fallback_location: documents_path
```
Comment (richrace) [inline app/controllers/documents_controller.rb]: Does using `ensure` work here so it's clear this should always happen?

```ruby
def review_failure
    @document = use_cases.review_document_failure.execute(document_id: params.require(:id))
    flash[:notice] = 'Successfully marked as reviewed'
  rescue Exceptions::IncomeApiError::NotFoundError
    flash[:notice] = "An error occurred while marking document #{params.require(:id)} as reviewed"
  ensure
    redirect_back fallback_location: documents_path
end 
```

File: spec/controllers/documents_controller_spec.rb [hunk 6]
```diff
@@ -98,4 +98,34 @@
...
+        patch :review_failure, params: { id: id }
+
+        expect(flash[:notice]).to eq('Successfully marked as reviewed')
+      end
+    end
+
+    context 'when not successfully reviewed' do
+      let(:document_response) { Net::HTTPResponse.new(1.1, 400, 'NOT OK') }
+
+      it 'should show a success message' do
```
Comment (alex-demetriou) [inline spec/controllers/documents_controller_spec.rb]:
```suggestion
      it 'should show an error message' do
```

Comment (Miles-Alford) [inline spec/controllers/documents_controller_spec.rb]: Shouldn't this test name be "It should show an error message"

**Pull request merged**